Work on show registrar users on the management site

maael opened the pull request.

maael committed:
Move RegiUser to Common project

maael committed:
Add basics for user sections of management site

maael committed:
Add UserGetAll query to registrar queries

maael committed:
Add GetUsers method to RegiUserStore

Also change other user get methods to use private GetUsersByQuery method
to reduce code duplication.

maael committed:
Add relay controller to management api to registrar api

Also add method to registrar connection to get list of users from
endpoint.

maael committed:
Add endpoints to registrar to get users list and details

maael committed:
Merge branch 'master' into bug/fix-vote-create-questions

maael committed:
Change user components to use component composition

maael committed:
Fix RegistrarConnectionController to use await correctly

maael committed:
Fix component capitalisation

maael committed:
Start work on user components

maael committed:
Fix RegiUser namespace, tidy using statements

maael committed:
Add response model of registrar user for the management site

Also -
- Add emailConfirmed fields to SingleRegiUserResponse models.
- Update RegistrarConnection.

maael committed:
Change Registrar User controller to use ApiKeyAuth from Authorize

maael committed:
Add management SingleRegiUserResponse model to project

maael committed:
Update user react components

MetalMichael commented on EvotoApi/Areas/Management/Connections/RegistrarConnection.cs:
```diff
@@ -15,18 +20,38 @@ public class RegistrarConnection
+        private static async Task<IRestResponse> MakeApiRequest(RestRequest req)
         {
             var client = new RestClient(RegistrarUrl);
+            req.AddQueryParameter("key", ConfigurationManager.AppSettings.Get("ApiKeys"));
```
Might consider checking this exists first, to save a null reference exception or maybe something else more hard to find

MetalMichael commented on EvotoApi/Areas/Management/ManagementApiAreaRegistration.cs:
```diff
@@ -19,6 +19,12 @@ public override void RegisterArea(AreaRegistrationContext context)
                 new { action = "Index", id = UrlParameter.Optional }
             );
+
+            context.MapRoute(
```
Why is this being defined dynamically?

maael replied:
The route wouldn't work without this, and I have minimal knowledge of how all the routing is handled on this project - there seems to be multiple different approaches.

MetalMichael replied:
There should either be static routing (using route attributes [RoutePrefix("test")] and [Route("subTest")] )

Which should override any mapped routes. Then route mapping is used when static routes aren't defined or found for the current route. 

If there is an issue, it could be caused by the order of things in WebApiConfig. I tweaked it a bit in the registrar api.
[RegistrarApi.WebApiConfig](https://github.com/evoto-tech/EvotoApi/blob/master/RegistrarApi/App_Start/WebApiConfig.cs)
[ManagementApi.WebApiConfig](https://github.com/evoto-tech/EvotoApi/blob/master/EvotoApi/App_Start/WebApiConfig.cs)

Just leave it for now then, I'll take a look when I refactor the projects

MetalMichael replied:
Maybe tag it with `// TODO` if you're making any changes

MetalMichael commented on Registrar.Database/Stores/RegiSqlUserStore.cs:
```diff
@@ -42,30 +43,22 @@ public RegiSqlUserStore(string connectionString) : base(connectionString)
-                using (var connection = await GetConnectionAsync())
-                {
-                    var result = await connection.QueryAsync(RegistrarQueries.UserGetByEmail, new {Email = email});
+            var users = await GetUserByQuery(RegistrarQueries.UserGetAll, new {});
```
Creating a pointless new object? Null not supported? Probably should overload the getuser method

MetalMichael commented on Registrar.Database/Stores/RegiSqlUserStore.cs:
```diff
@@ -16,19 +18,18 @@ public RegiSqlUserStore(string connectionString) : base(connectionString)
+                    var result = await connection.QueryAsync(query, parameters);
 
                     if (!result.Any())
                         throw new RecordNotFoundException();
```
Probably should throw exception in public methods, in case one isn't wanted/needed. For example if selecting a list of all users, probably don't want an exception, just an empty list.

MetalMichael commented on RegistrarApi/Controllers/UserController.cs:
```diff
@@ -0,0 +1,51 @@
+        [ApiKeyAuth]
+        public async Task<IHttpActionResult> List ()
+        {
+            var details = await _store.GetUsers();
```
Catch exception? (see previous comment)

MetalMichael commented on RegistrarApi/Controllers/UserController.cs:
```diff
@@ -0,0 +1,51 @@
+        [ApiKeyAuth]
+        public async Task<IHttpActionResult> Details(int userId)
+        {
+            var details = await _store.GetUserById(userId);
```
try { ... } catch (RecordNotFoundException) { return NotFound(); }

MetalMichael commented on RegistrarApi/Controllers/UserController.cs:
```diff
@@ -0,0 +1,51 @@
+
+        #region Helpers
+
+        private void AddErrors(IdentityResult result)
```
Unused?

maael committed:
Fix with feedback from QA

- Add TODO note about regi api routing
- Add check for ApiKeys value in RegistrarConnection
- Add try/catches for UserController
- Remove unused Helpers from UserController
